utils/utf8.cc

```diff
@@ -56,22 +56,30 @@ namespace utils {
 
 namespace utf8 {
 
-// 3x faster than boost utf_to_utf
-static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t len) {
-    size_t pos = 0;
-
-    while (len) {
-        size_t bytes;
-        const uint8_t byte1 = data[0];
+struct codepoint_status {
+    size_t bytes_validated;
+    bool error;
+    uint8_t more_bytes_needed;
+};
 
+static
+codepoint_status
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
```

Comment:
std::array? and const?

Reply:
std::array gives nothing here IMO. It's useful to wrap an array in a regular object without special behavior when passing it around, especially to templates. Here it's just obfuscation.

Added const.

Reply:
> I think the validator was missing some important checks and didn't understand other checks it did - see my review.

This series does not introduce a validator, so comments about the validator do not apply.

> If any of the potential bugs I raised are true, it also means the tests missed them so they aren't exhaustive enough.
> Moreover, I think the fragmented utf8-checking should actually be simpler than the way you did it. If a UTF-8 "code point" is split between buffers, there is no reason to merge the two halfs. The first half includes the first byte - which indicates the code's **length**, and all the rest of the bytes simply need to begin by bits 10 - i.e., be between 0x80 and 0xBF. 

That's not true. 

Edit: actually it is true. So maybe it can be simplified.

>In other words, in the second buffer, we just need to check that the first M (calculated in the previous buffer) bytes are between 0x80 and 0xBF - there is no need to keep any byte data from the previous buffer.

Reply:
Edit2: it's not true.The sequence f0 80 80 80 codes for codepoint 0, but is illegal.

Reply:
I was under the impression that in modern C++ std::array should always replace old arrays when the size is known. Isn't std::array safer?

Reply:
No, it just wraps a regular C array.

It's safer when you pass it around, since C arrays have strange rules where they decay into a pointer and lose information about the size.

Reply:
Fair enough. It's not safer. It still seems more idiomatic in modern c++. Because why would anyone use C arrays when there is std::array and std::vector?

Reply:
> Edit2: it's not true.The sequence f0 80 80 80 codes for codepoint 0, but is illegal.

I didn't think about this. But not all is lost. Basically, leading zeros are not allowed. So it's not enough to remember remaining_bytes for the current codepoint, we also need to remember is_second_byte which means whether the next byte to be read is the first one between 0x80 and 0xBF - and if it is the first one, it must not be 0x80.

Anyway, I see now that https://tools.ietf.org/html/rfc3629 has in "Syntax of UTF-8 Byte Sequences" the same weird ranges that I don't understand. Maybe I'm missing something important, but I don't understand yet what :-(

Reply:
Yes, but I don't want to invent a new mechanism for validating the buffer-straddling codepoint. There's no real benefit, and there is a risk of getting the validation wrong.